File: 0202-happy-number/0202-happy-number.java
import java.util.HashSet;

class Solution {

    public int square_digits(int num) {
        int sum = 0;
        while (num > 0) {
            int digit = num % 10;
            sum += digit * digit;
            num /= 10;
        }
        return sum;
    }
    public boolean isHappy(int n) {

        HashSet<Integer> patterns = new HashSet<>();

        int next_num = n;

        while (!patterns.contains(next_num)) {
            patterns.add(next_num);
            next_num = square_digits(next_num);
            if (next_num == 1) return true;
        } 

        return false;
    }
}

File: 0202-happy-number/0202-happy-number.py
class Solution:

    def square_digit(self, n: int) -> int:
        final_sum = 0
        while n > 0:
            digit = n % 10
            n = n // 10
            final_sum += digit * digit
        return final_sum


    def isHappy(self, n: int) -> bool:

        fast = n
        slow = n

        while (True):
            slow = self.square_digit(slow)
            fast = self.square_digit(self.square_digit(fast))

            if (fast == 1): return True
            if (fast == slow): return False
    
        